src/HieDb/Utils.hs

```diff
@@ -38,25 +38,39 @@ import Data.IORef
 import HieDb.Types
 import HieDb.Compat
 import Database.SQLite.Simple
+import Control.Monad.State.Strict (StateT, get, put)
+import Data.Set (Set)
+import qualified Data.Set as Set
+import Control.Monad (when)
 
 #if __GLASGOW_HASKELL__ >= 903
 import Control.Concurrent.MVar (readMVar)
 #endif
 
-addTypeRef :: HieDb -> FilePath -> A.Array TypeIndex HieTypeFlat -> A.Array TypeIndex (Maybe Int64) -> RealSrcSpan -> TypeIndex -> IO ()
+-- Each AST Node can have a lot of repetitive type information.
+-- We use this to make sure that each occurrence of a type
+-- (identified by Int64 id of that type within typenames table)
+-- is inserted exactly once per each RealSrcSpan in which it occurs and per
+-- depth (Int) of that type within the tree structure representing the type
+type TypeIndexing a = StateT (Set (Int64, Int)) IO a
```

Comment:
could also be a `IntMap IntSet` (morally `Map Depth (Set TypeIndex)`) as a microoptimisation

Reply:
I won't block this from merging if you don't feel like it though.

Reply:
Good call. Applied and this improved the indexing time of by another ~20-30% :smile: